tao_compiler/mlir/disc/transforms/disc_memref_gvn.cc

```diff
@@ -0,0 +1,144 @@
+#include "mlir/IR/Location.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "placement_utils.h"
+#include "tensorflow/compiler/mlir/disc/disc_util.h"
+#include "tensorflow/compiler/mlir/disc/transforms/PassDetail.h"
+
+namespace mlir {
+namespace disc_ral {
+
+namespace {
+
+struct LoadScalarSimplifier : public OpRewritePattern<memref::LoadOp> {
+ public:
+  LoadScalarSimplifier(MLIRContext* context, DominanceInfo* dominance_info)
+      : OpRewritePattern<memref::LoadOp>::OpRewritePattern(context),
+        dominance_info_(dominance_info) {}
+
+  LogicalResult matchAndRewrite(memref::LoadOp op,
+                                PatternRewriter& rewriter) const override;
+
+ private:
+  bool extractScalarIndex(ValueRange indices, int64_t& index) const;
+  DominanceInfo* dominance_info_;
+};
+
+bool LoadScalarSimplifier::extractScalarIndex(ValueRange indices,
+                                              int64_t& index) const {
+  if (indices.size() != 1) {
+    return false;
+  }
+  auto op = indices[0].getDefiningOp();
+  if (auto constOp = dyn_cast_or_null<arith::ConstantIndexOp>(op)) {
+    index = constOp.value();
+    return true;
+  }
+  return false;
+}
+
+// Replace the pattern of:
+//   memref.store %value, %memref_x[%index_x] : memref<?xtype>
+//   %a = memref.load %memref_x[%index_x] : memref<?xtype>
+//   Use of `%a`
+// with:
+//   memref.store %value, %memref_x[%index_x] : memref<?xtype>
+//   Use of `%value`
+// The store op could be eliminated in other passes (e.g., DCE).
+// Note: we only deal with memrefs on CPU in this pass currently. Meanwhile,
+// we only deal with the case that the loaded memref are only stored once at
+// the same index.
+LogicalResult LoadScalarSimplifier::matchAndRewrite(
+    memref::LoadOp op, PatternRewriter& rewriter) const {
+  Location loc = op.getLoc();
+  int64_t load_index;
+  if (!extractScalarIndex(op.getIndices(), load_index)) {
+    return failure();
+  }
+
+  Value load_memref = op.getMemRef();
+  if (placement_utils::isGpuMemRef(load_memref)) {
+    return failure();
+  }
+
+  memref::StoreOp store_op;
+  for (Operation* user : load_memref.getUsers()) {
+    // Before check alias, we first check dominance from `op` to `user`, which
+    // helps to reduce the cases of failure caused by alias dominated by `op`.
+    if (dominance_info_->dominates(op.getOperation(), user)) {
+      continue;
+    }
+    if (IsMemRefAliasOp(user) ||
+        (!isa<memref::StoreOp>(user) && IsOpWriteValue(user, load_memref))) {
+      // To prevent indirect modification on `load_memref`.
+      return failure();
+    }
+    // Only store-ops dominate `op` will be considered for optimization.
+    if (!isa<memref::StoreOp>(user) ||
+        !dominance_info_->dominates(user, op.getOperation())) {
```

Comment:
a store op user of a value can not be safely skipped when the store does not dominants the load op.

```
store  // a
if (%pred) {
  store // b
}
load // c
```  
As shown in the above example, `b` can not be safely ignored.

Reply:
To fix.